### DISCUSSION TITLE
Negative values of bands at level2

### DISCUSSION THREAD

**Discussion by cxxuuuuuuu** [davidfrantz/force#134] [Q&A] [answered]
Dear all,

have you noticed that some bands of Sentinel-2 BOA images have negative values after the  level2 processing? It will definitely affect the band math. 

![FORCE_forum](https://user-images.githubusercontent.com/84906020/138107141-dfec7bec-a82a-4f97-93ee-f473241fac8d.png)

 
Is it a quality issue of level 1 data? I would appreciate hearing your opinions on this. Thanks in advance;)

Cheers,
Chunyan

**Comment by davidfrantz** [comment]
Hi @cxxuuuuuuu 

in principle, reflectance has a physical limit of [0-1]. 
However, in radiometric correction, we are treating it as quantitative variable. There are additive effects, for which we need to correct.

Most notably, this is the path radiance in the atmospheric correction. Thus if AOD is slightly overestimated, you will get negative values in short wavelengths (most pronounced in blue).

When red or NIR are negative, this mostly happens due to the environment correction, where we essentially substract the background contribution (because photons bounce from neighboring surfaces onto our target). Negative values happen mostly on dark surfaces, e.g. tree or building shadows. We have no way of knowing where these occur (as opposed to terrain shadows) and the main problem here is the following: In atmospheric correction, we are correcting for the double atmospheric path (sun to target + target to sensor). However, in reality we only have one path (target to sensor because the incoming light was blocked + some diffuse illumination).

Note that the TSA module in FORCE Higher Level makes it easy for you to account for this. First, it can handle all the quality masking (there is a subzero flag in the ``QAI`` product), and secondly, it safeguards you from negative (or super-positive) NDVI values (there is a check implemented).

Hope this helps somewhat,
David

**Reply by cxxuuuuuuu** [reply]
Hi David,

thanks for your reply. I will use the TSA module to avoid the negative (or super-positive) values, and come back here to tell you if it works.

Best regards,
Chunyan

**Reply by cxxuuuuuuu** [reply]
Hi @davidfrantz 

now I am using TSA to calculate some vegetation indices (NDVI, NBR etc.). But I couldn't generate any results. After running the TSA parameter file (tsi.prm), I always get the same feedback like:

eouser@force:~$ force-higher-level /home/eouser/src/force/param/tsi.prm
![1_TSA_1](https://user-images.githubusercontent.com/84906020/140745565-70b78219-1c6e-45da-b8d7-60a6f3aa7678.png)


I have noted that all the 'time' are 00, which is definitely wrong, but I don't know why.

Here is my parameter file.
[tsi.txt](https://github.com/davidfrantz/force/files/7496941/tsi.txt)

My Level-2 ARD images are under:
![1_TSA](https://user-images.githubusercontent.com/84906020/140745183-bd5c7ee1-c504-440d-a65a-c1107b50e7d6.png)


And the mask raster is under 
![2_TSA](https://user-images.githubusercontent.com/84906020/140745217-b883efaa-f057-4585-8dfc-88d3bdecbd8c.png)


![3_TSA](https://user-images.githubusercontent.com/84906020/140745236-03685278-8d45-4f9c-907b-65a7aa249232.png)


Did I make obvious mistakes at storing the mask or Level 2 ARD images? 

I would appreciate hearing your opinions on this. Thanks a lot !!


Best regards,
Chunyan

**Reply by davidfrantz** [reply]
Hi @cxxuuuuuuu,

yes, there is a mistake in l. 10:

```
DIR_MASK = /home/eouser/src/force/TSA/mask/X0000_Y0000 
-->
DIR_MASK = /home/eouser/src/force/TSA/mask
```

Best,
David

**Reply by cxxuuuuuuu** [reply]
Hi @davidfrantz 

thanks a lot for your quick reply. I changed the path of mask raster to DIR_MASK = /home/eouser/src/force/TSA/mask :

![1_TSA_2](https://user-images.githubusercontent.com/84906020/140758733-51cc856f-22e8-49e5-9dfb-9bb63ff2c867.png)

But the feedback of console is still the same:

![1_TSA_3](https://user-images.githubusercontent.com/84906020/140759189-a37a1a7d-27a6-4b89-809e-0b35c185312c.png)

So I guess the problem is somewhere else... Really complex.


Best regards,
Chunyan

**Comment by davidfrantz** [comment, marked as answer]
ehm, did you move the tif one folder up?

The masks needs to be in the tiles.

When processing, FORCE will append the directory name with the current tile, e.g. when the current tile is 0/0, 
it will automatically look in 

``/home/eouser/src/force/TSA/mask/X0000_Y0000``

Best,
David

**Reply by cxxuuuuuuu** [reply]
Hi @davidfrantz 

thanks for your explanation!! I had misunderstanding that I should change the path of mask raster. 

Now it works:

![5_TSA](https://user-images.githubusercontent.com/84906020/140806811-714fe927-9b04-4260-a100-f905d80ec9ea.png)

I will take a look at the interpolated results. If they look good, I would tell you and you can mark this discussion as "answered" ;)


Best regards and thanks again!
Chunyan

**Comment by cxxuuuuuuu** [comment]
Hi @davidfrantz 

the TSS rasters look very good and they don't have extreme negative (or super-positive) values anymore. Thank you again for your patient explanation and replies.

Besides, I am just wondering whether I can calculate more other vegetation indices (VIs) with TSA module? Because in your tutorial there is a limited list of "Spectral index" (https://force-eo.readthedocs.io/en/latest/components/higher-level/tsa/param.html). What if I would like to calculate VIs which are not in the list of "Spectral index"? Is it possible to use PYTHON-PLUGIN PARAMETERS?

I am looking forward to your replies and greetings from Berlin.


Best regards,
Chunyan

**Comment by davidfrantz** [comment]
You're welcome!

There are quite some indices built into FORCE. 
If you want to have other indices, you need to fork the repository, change the code, and then Pull-Request to link the changes back.

Or, as you mentioned, you can use the UDF submodule and compute via Python code snippets.

Please see this tutorial here: https://force-eo.readthedocs.io/en/latest/howto/udf.html

Best,
David
